src/graph/paths.cc

```diff
@@ -918,7 +918,11 @@ ncclResult_t ncclTopoComputeP2pChannels(struct ncclComm* comm) {
   // fill the whole space of nChannels. To do so we mirror the bits in the
   // nChannels space.
   for (int c=0; c<comm->p2pnChannels; c++) {
-    comm->p2pChannels[c] = mirrorBits(c, comm->p2pnChannels);
+    //update mirrorbits to support more than 64 p2p channels
+    if (comm->p2pnChannels == 2*CHANNEL_LIMIT)
+      comm->p2pChannels[c] = mirrorBits(c, comm->p2pnChannels);
+    else
+      comm->p2pChannels[c] = mirrorBits(c, max(64,comm->p2pnChannels/2));
```

Comment:
I made a mirrorBits test app. it actually works fine to 128 channels. The logic here is wrong instead.

#include <stdio.h>

static int mirrorBits(int val, int pow2) {
  int mirror = 0;
  for (int b=1, mb=(pow2>>1); b<pow2; b<<=1, mb>>=1) if (val & b) mirror |= mb;
  return mirror;
}


int main()  {
	int nchans = 128;
	int mirrors[nchans];
	for (int c = 0; c < nchans; c ++) {
		mirrors[c] = mirrorBits(c, nchans);
		printf("c %d mirror %d\n", c, mirrors[c]);
	}
	for (int c = 0; c < nchans; c ++) {
		for (int d = 0; d < nchans; d ++) {
			if (c == d) continue;
			if (mirrors[c] == mirrors[d])
				printf("Error! Duplication detected at %d and %d\n", c, d);
		}
    }
    printf("All check passed!\n");
}

Reply:
Does not work for RCCL pt2pt tests.

Reply:
so p2pnchannels cannot be more than 64 for now? if this is the case, we should limit it to 64 at the place where it is calculated. Not initializing the channels but not use it somehow